x-pack/plugin/spatial/build.gradle

```diff
@@ -11,6 +11,7 @@ esplugin {
 }
 
 dependencies {
+  api project(path: ':modules:legacy-geo')
```

Comment:
Just to be clear, this will cause a copy of the legacy-geo jar (and it's dependencies) to be added to the xpack spatial module. It won't reuse the same. I think that is ok, as there is no communication between them, I believe this is just for reuse of some common stuff, but I want to double check with you of that understanding.

Reply:
We need this because when creating a geo_shape field mapper we might need to create the mapper on the legacy jar: 

https://github.com/elastic/elasticsearch/blob/2e25028a479facd85e1fdc3b0657e68dc00172bf/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java#L205

Reply:
We need this because when creating a geo_shape field mapper we might need to create the mapper on the legacy jar: 

https://github.com/elastic/elasticsearch/blob/2e25028a479facd85e1fdc3b0657e68dc00172bf/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java#L205

Reply:
Is that the only use? If so, consider moving the shared code into a project under lib. By depending on the module the dependencies, including jts, are pulled into (as a copy) the spatial module.

Reply:
I think it is tricky, the mapper and shape builders companions are tightly dependent on Elasticsearch server and other projects under lib.

Reply:
Ok, it's critical to move the common stuff out. And as long as there isn't any *interaction* between the legacy module and spatial module, then there isn't a problem. But if, eg spatial grabs a FieldMapper and tries to cast it to a class that is from the legacy module, it will cause problems because it is a different Class instance, even though it's the same source code. There are ways around this, but I wanted to be sure it was necessary before suggesting anything.

Reply:
Field mappers have this underlaying method:

https://github.com/elastic/elasticsearch/blob/b2ed599ec5fa09d6358c9f8cf61ca778321d510f/server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java#L356

Would that be an issue?

Reply:
No, I don’t think that is an issue. The mapper is checking the existing field type against itself.

Reply:
I think there is no interaction between the modules except for the spatial module creating Legacy mappers under certain circumstances. Still I am thinking if we can do better without being too invasive.

Reply:
How does spatial create the legacy mappers, through normal mapper apis (ie json) or directly in code? If it is in code, then I don't think that would work.

The "solution" is to do something we recommend against in general, which is declaring the spatial module "extends" legacy geo. Even though it doesn't actually extend it, it will make the classloader of legacy geo a parent of spatial, so the classes would be available. To do this you would add the new module to the extendedPlugins list in build.gradle, and change the dependencies section to use compileOnly instead of api for the legacy module.

Reply:
I implemented the "solution" above, therefore now spatial module extends legacy module and dependency
is compile only now. It was very insightful and I got a better understanding how it works. This action has some side effects:

- Vector tile extends the spatial module but now in order to compile I need to add a compile only dependency to JTS
- Couple of testing modules were adding spatial as a runtime dependency. I have to add legacy-geo as well to make it happy.

Reply:
@rjernst Do you think this is good enough or we should have more iterations?